File: Top Interview Questions/Linked List/206. Reverse Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        """
        # recursive O_n time and O_n space
        if not head or not head.next:
            return head
        
        p = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        
        return p
        """
        
        # iterative O_n time and O_1 space
        prev, curr = None, head
        
        while curr:
            temp = curr.next
            curr.next = prev
            prev, curr = curr, temp
            
        return prev
    

File: Top Interview Questions/Linked List/206. Reverse Linked List.java
/**
 * 206. Reverse Linked List
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        // stack approach O_n time and O_n space
        if (head == null || head.next == null) {
            return head;
        }

        Stack<Integer> s = new Stack<Integer>();

        while (head != null) {
            s.push(head.val);
            head = head.next;
        }

        ListNode a = new ListNode(0);
        ListNode dummy = a;

        while (!s.empty()) {
            a.next = new ListNode(s.pop());
            a = a.next;
        }

        return dummy.next;

    }
}